contracts/service-registry/src/contract.rs

```diff
@@ -100,60 +136,125 @@ pub mod execute {
+                |sw| -> Result<Worker, ContractError> {
+                    match sw {
+                        Some(worker) => Ok(Worker {
+                            authorization_state: AuthorizationState::NotAuthorized,
+                            ..worker
+                        }),
+                        None => Ok(Worker {
+                            address: worker,
+                            bonding_state: BondingState::Unbonded,
+                            authorization_state: AuthorizationState::NotAuthorized,
+                            service_name: service_name.clone(),
+                        }),
+                    }
+                },
+            )?;
         }
 
-        let worker_address = info.sender.clone();
+        Ok(Response::new())
+    }
+
+    pub fn bond_worker(
+        deps: DepsMut,
+        info: MessageInfo,
+        service_name: String,
+    ) -> Result<Response, ContractError> {
+        SERVICES
+            .may_load(deps.storage, &service_name)?
+            .ok_or(ContractError::ServiceNotFound {})?;
+
+        let bond = if !info.funds.is_empty() {
+            info.funds
+                .iter()
+                .find(|coin| coin.denom == AXL_DENOMINATION)
+                .ok_or(ContractError::WrongDenom {})?
+                .amount
+        } else {
+            Uint128::zero() // sender can rebond currently unbonding funds by just sending no new funds
+        };
 
-        service_workers().update(
+        WORKERS.update(
             deps.storage,
-            &info.sender,
+            (&service_name.clone(), &info.sender.clone()),
             |sw| -> Result<Worker, ContractError> {
                 match sw {
-                    Some(found) => {
-                        if found.state == WorkerState::Inactive {
-                            Ok(Worker {
-                                address: worker_address,
-                                stake: bond.unwrap().amount,
-                                commission_rate,
-                                state: WorkerState::Active,
-                                service_name,
-                            })
-                        } else {
-                            Err(ContractError::ServiceWorkerAlreadyRegistered {})
-                        }
+                    Some(worker) => {
+                        let bonding_state = match worker.bonding_state {
+                            BondingState::Bonded { amount }
+                            | BondingState::RequestedUnbonding { amount }
+                            | BondingState::Unbonding {
+                                amount,
+                                unbonded_at: _,
+                            } => BondingState::Bonded {
+                                amount: amount + bond,
+                            },
+                            BondingState::Unbonded {} => BondingState::Bonded { amount: bond },
+                        };
```

Comment:
extract into a method of `BondingState`? `fn add_bond(self) -> Self` It can be easily unit-tested as well that way.